### PR TITLE
Module for Spark Integration with Druid.

### PR DESCRIPTION
**Pull request by deepikakhera**
Added a Module for Integration Spark Applications with Druid.

### CODE REVIEW

**Comment by xvrl** [comment]
thanks for the contribution @deepikakhera!
it looks like there are some jackson-scala version conflicts in the scala 2.11 build:

`java.lang.RuntimeException: Conflicting cross-version suffixes in: com.fasterxml.jackson.module:jackson-module-scala`

You can see the full build output here https://travis-ci.org/druid-io/tranquility/jobs/94716045

**Comment by deepikakhera** [comment]
Thanks Xavier. Yes, I am looking at it now.

File: build.sbt [hunk 1]
```diff
@@ -78,10 +79,18 @@ val samzaDependencies = Seq(
   "org.apache.samza" % "samza-api" % samzaVersion % "optional"
 )
 
+val sparkDependencies = Seq (
+  "org.apache.spark" % "spark-streaming_2.10" % sparkVersion % "optional"
```
Comment (gianm) [inline build.sbt]: Perhaps try:

```scala
"org.apache.spark" %% "spark-streaming" % sparkVersion % "optional"
```

The %% will make sbt pull in the appropriate 2.10 or 2.11 version depending on what we're building with.

File: build.sbt [hunk 2]
```diff
@@ -18,6 +18,7 @@ val druidVersion = "0.8.1"
 val finagleVersion = "6.25.0"
 val twitterUtilVersion = "6.25.0"
 val samzaVersion = "0.8.0"
+val sparkVersion = "1.4.1"
```
Comment (gianm) [inline build.sbt]: I think Spark is on 1.5 now, could we use that or is there an issue preventing it?

File: README.md [hunk 3]
```diff
@@ -276,6 +276,61 @@ class MyBeamFactory extends BeamFactory[Map[String, Any]]
...
+If you're using Spark to process your event stream, you can use Tranquility's BeamRDD adapter to send data to
+Druid. The BeamRDD class provides any RDD the ability to write to Druid using the propagate() function. 
+The BeamRDD requires a BeamFactory to propagate events. You will be overriding the makeBeam() function and within that using the DruidBeams builder's "buildBeam()" to build the beam. 
+It is recommended that you implement makeBeam as a lazy val. 
+
+For example:
+
+```scala
+
+class SimpleEventBeam extends BeamFactory[SimpleEvent] {
```
Comment (gianm) [inline README.md]: Please rename to `SimpleEventBeamFactory`, as this is a BeamFactory not a Beam

File: build.sbt [hunk 4]
```diff
@@ -155,6 +164,13 @@ lazy val core = project.in(file("core"))
   .settings(publishArtifact in(Test, packageBin) := true)
   .settings(libraryDependencies ++= (coreDependencies ++ coreTestDependencies))
 
+lazy val spark = project.in(file("spark"))
+  .settings(commonSettings: _*)
+  .settings(name := "tranquility-spark")
+  .settings(resolvers += "clojars" at "http://clojars.org/repo/")
```
Comment (gianm) [inline build.sbt]: I think you don't need the clojars repo here. It's needed for clojure libs in storm, but I don't think spark uses those.

File: spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  * @param rdd rdd that should be propagated to druid.
+  */
+class BeamRDD[T: ClassTag](rdd:RDD[T]) extends Logging with Serializable{
+
+  def propagate(beamFactory: BeamFactory[T]) = {
+    rdd.foreachPartition {
+      partitionOfRecords => {
+        val beam = beamFactory.makeBeam
+        val eventsWritten = Await.result(beam.propagate(partitionOfRecords.toIndexedSeq))
+        log.info("Propagated "+ eventsWritten + " events to Druid ")
```
Comment (gianm) [inline spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala]: (minor) You can use string interpolation and do `s"Propagated $eventsWritten events to Druid."`

File: spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import com.metamx.common.scala.Logging
+import scala.language.implicitConversions
+
+/**
+  * This class provides any RDD the ability to propagate events to Druid.
+  * @param rdd rdd that should be propagated to druid.
+  */
+class BeamRDD[T: ClassTag](rdd:RDD[T]) extends Logging with Serializable{
+
+  def propagate(beamFactory: BeamFactory[T]) = {
```
Comment (gianm) [inline spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala]: Hmm, maybe we should call this something else that makes it more clear to readers of the spark code what's going on. `sendWithTranquility`?

Comment (gianm) [inline spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala]: `propagate` is ok too I guess, the BeamFactory makes it pretty clear what's going on.

Comment (deepikakhera) [inline spark/src/main/scala/com/metamx/tranquility/spark/BeamRDD.scala]: Thanks a ton for the feedback Gian! I am working on them, will check in the
changes in a little bit.

Best,
Deepika

**Comment by gianm** [comment]
@deepikakhera could you please download the style jar at https://github.com/druid-io/druid/blob/master/intellij_formatting.jar and apply that? You can import it into IntelliJ and select the "Druid Java and Scala style" code style.

**Comment by deepikakhera** [comment]
Sure, will do. Thanks Gian.

**Comment by gianm** [comment]
@deepikakhera could you also please fill out the Druid CLA at http://druid.io/community/cla.html? It applies to tranquility as well.

Thanks!

**Comment by deepikakhera** [comment]
Hi,

Have made the recommended changes.
The Integration tests are passing now. Please take a look.

Thanks,
Deepika

**Comment by gianm** [comment]
@deepikakhera this looks good to me, could you please squash your commits? You could follow the instructions on https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md.

**Comment by deepikakhera** [comment]
Cool, thanks Gian. Yes, will do that.

**Comment by deepikakhera** [comment]
Hi Gian,

I have squashed my commits and pushed my changes. Please let me know if
they look ok now.

Thanks,
Deepika

**Comment by gianm** [comment]
@deepikakhera looks great, :+1:

thanks!

**Comment by deepikakhera** [comment]
Cool, so is there any other action required on my side?

**Comment by gianm** [comment]
@deepikakhera not right now, just need someone else to take a look and merge.

**Comment by deepikakhera** [comment]
okie, thanks ! Will stand by then.

**Comment by fjy** [comment]
:+1:

**Comment by deepikakhera** [comment]
Thanks for merging the changes Fangjin!

**Pull request merged**